pkg/forward/ringbuffer.go

```diff
@@ -0,0 +1,87 @@
+package forward
+
+import (
+	"sync"
+)
+
+// RingBuffer is a fixed-length ring buffer. It can be used by a forwarder, to 'drop' messages instead of applying backpressure. See Issue #15
+type RingBuffer struct {
+	max int
+
+	buf   []string
+	first int          // the index of the first item in the buffer
+	last  int          // the index of the last item in the buffer
+	ch    chan string  // the channel is used to block on the Get method whenever data is unavailable
+	mutex sync.RWMutex // synchronizes changes to buf, first, last
+}
+
+// Put() processes the record without blocking.
+// It's behaviour varies depending on the state of the buffer and any blocking Get() invocations
+// It either sends the record over the channel, adds it to the buffer, or drops the record if the buffer is full.
+func (b *RingBuffer) Put(record string) {
+	select {
+	case b.ch <- record: // no need to even add to the buffer if something is blocking on the channel
+	default:
+		b.mutex.Lock()
+		defer b.mutex.Unlock()
+		if b.len() >= b.max {
+			// Drop record
+			// Note: another alternative would have been to move the 'first' value before putting this record onto the buffer, but it's easier and marginally cheaper to just drop the incoming record.
+			return
+		}
```

Comment:
Ringbuffers should always cache the most recent data, so yeah, you'll want to drop the oldest record and keep the incoming record.